Thirty/Thirty/Utilities/FirebaseManager.swift

```diff
@@ -247,66 +182,69 @@ class FirebaseManager {
-            completion(.Failure(error))
+            completion(.failure(error))
         }
     }
+
+    // MARK: Friends
     
     func addUserAsFriend(username: String, completion: @escaping (Result<Void>) -> Void) {
-        databaseRef.child("friends")
+        friendsRef
             .child(UserManager.shared.currentUserUsername.lowercased())
             .setValue([username.lowercased(): true], withCompletionBlock: { [weak self] (error, ref) in
                         if let error = error {
-                            completion(.Failure(error))
+                            completion(.failure(error))
                         } else {
                             self?.databaseRef.child("friends")
                                 .child(username.lowercased())
                                 .setValue([UserManager.shared.currentUserUsername.lowercased(): true], withCompletionBlock: { (error, ref) in
                                     if let error = error {
-                                        completion(.Failure(error))
+                                        completion(.failure(error))
                                     } else {
-                                        completion(.Success)
+                                        completion(.success)
                                     }
                                 })
                         }
             })
     }
-    
+
     func getContacts(completion: @escaping (Result<Void>) -> Void) {
-        databaseRef.child("friends").child(UserManager.shared.currentUserUsername.lowercased()).observeSingleEvent(of: .value, with: { [weak self] snapshot in
-            if let value = snapshot.value as? NSDictionary,
-                let usernames = value.allKeys as? [String] {
-                let dispatchGroup = DispatchGroup()
-                for username in usernames {
-                    dispatchGroup.enter()
-                    self?.databaseRef.child("users").child(username).observeSingleEvent(of: .value) { snapshot in
-                        let value = snapshot.value as! [String: Any]
-                        let user = User()
-                        user.username = value["display-name"] as! String
-                        user.deviceToken = value["device-token"] as? String
-                        user.email = value["email"] as! String
-                        user.phoneNumber = value["phone-number"] as! String
-                        UserManager.shared.contacts.append(user)
-                        dispatchGroup.leave()
-                    }
-                }
-                dispatchGroup.notify(queue: DispatchQueue.global(qos: .`default`)) {
-                    DispatchQueue.main.async {
-                        UserManager.shared.contacts.sort(by: { $0.username.lowercased() < $1.username.lowercased() })
-                        completion(.Success)
-                    }
+        let cancelBlock: (Error) -> Void = { completion(.failure($0) )}
```

Comment:
the firebase cancelBlock has me a little confused - what's the purpose/when would it get called?  It's not called if there's a firebase error, correct?

Reply:
It only gets fired when the user doesnt have _permission_ to perform that request/access that data. Basically if firebase deems your action "illegal" and thus cancels your action